packager/packager.cc

```diff
@@ -0,0 +1,731 @@
+    return ast::MK::UnresolvedConstant(core::LocOffsets::none(), move(scope), name);
+}
+
+ast::TreePtr PackageName::getFullNameLiteral(core::LocOffsets loc) const {
+    ast::TreePtr name = ast::MK::EmptyTree();
+    for (auto namePart : fullName) {
+        name = name2Expr(namePart, move(name));
+    }
+    // Outer name should have the provided loc.
+    if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(name)) {
+        name = ast::MK::UnresolvedConstant(loc, move(lit->scope), lit->cnst);
+    }
+    return name;
+}
+
+ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameRef fun, ast::TreePtr &expr) {
+    auto target = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    if (target == nullptr) {
+        if (auto e = ctx.beginError(expr->loc, core::errors::Packager::InvalidImportOrExport)) {
+            e.setHeader("Argument to `{}` must be a constant", fun.show(ctx));
+        }
+    }
+    return target;
+}
+
+struct PackageInfoFinder {
+    unique_ptr<PackageInfo> info = nullptr;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exported;
+    vector<unique_ptr<ast::UnresolvedConstantLit>> exportedMethods;
+    core::LocOffsets exportMethodsLoc = core::LocOffsets::none();
+
+    ast::TreePtr postTransformSend(core::MutableContext ctx, ast::TreePtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+
+        // Ignore methods
+        if (send.fun == core::Names::keepDef() || send.fun == core::Names::keepSelfDef()) {
+            return tree;
+        }
+
+        // Blacklisted methods
+        if (send.fun == core::Names::extend() || send.fun == core::Names::include()) {
+            if (auto e = ctx.beginError(send.loc, core::errors::Packager::InvalidPackageExpression)) {
+                e.setHeader("Invalid expression in package: `{}` is not allowed", send.fun.data(ctx)->shortName(ctx));
+            }
+            return tree;
+        }
+
+        // Sanity check arguments for unrecognized methods
+        if (send.fun != core::Names::export_() && send.fun != core::Names::import() &&
+            send.fun != core::Names::exportMethods()) {
+            for (const auto &arg : send.args) {
+                if (!ast::isa_tree<ast::Literal>(arg)) {
+                    if (auto e = ctx.beginError(arg->loc, core::errors::Packager::InvalidPackageExpression)) {
+                        e.setHeader("Invalid expression in package: Arguments to functions must be literals");
+                    }
+                }
+            }
+        }
+
+        if (info == nullptr) {
+            // We haven't yet entered the package class.
+            return tree;
+        }
+
+        if (send.fun == core::Names::export_() && send.args.size() == 1) {
+            // null indicates an invalid export.
+            if (auto target = verifyConstant(ctx, core::Names::export_(), send.args[0])) {
+                exported.emplace_back(copyConstantLit(target));
```

Comment:
Do I understand this right that you copy constant literal trees from one file, and then paste them into other files, but don't change all of their LocOffsets?

This is a recipe for a crash. Create a really big `__package.rb` file (it could be nothing but comments honestly, plus one export) and then try to import that package into another, really small `__package.rb` file. The LocOffset will be a begin/end offset against the source text of the file they were pasted into, but since the first file was bigger than the second file, you'll get an off-the-end read of the buffer, which should probably crash Sorbet if you ever manage to get that loc to be used (in error reporting or something else).

Reply:
Good point. I can change this logic to scrub the locs.

Reply:
Fixed.